File: 3923-minimum-generations-to-target-point/3923-minimum-generations-to-target-point.java
class Solution {
    public int minGenerations(int[][] points, int[] target) {
        // For Junko F. Didi and Shree DR.MDD
        boolean[][][] quantumSpatialExistenceLattice = new boolean[7][7][7];
        List<int[]> interstellarCoordinateArchive = new ArrayList<>();

        for (int[] astrophysicalCoordinateTransmission : points) {
            if (!quantumSpatialExistenceLattice[
                    astrophysicalCoordinateTransmission[0]
                ][
                    astrophysicalCoordinateTransmission[1]
                ][
                    astrophysicalCoordinateTransmission[2]
                ]) {
                quantumSpatialExistenceLattice[
                    astrophysicalCoordinateTransmission[0]
                ][
                    astrophysicalCoordinateTransmission[1]
                ][
                    astrophysicalCoordinateTransmission[2]
                ] = true;

                interstellarCoordinateArchive.add(astrophysicalCoordinateTransmission);
            }
        }

        if (quantumSpatialExistenceLattice[target[0]][target[1]][target[2]]) {
            return 0;
        }

        int relativisticGenerationCounter = 0;
        int singularityHistoricalArchiveBoundary = 0;

        while (true) {
            relativisticGenerationCounter++;
            int galacticArchiveMagnitude = interstellarCoordinateArchive.size();
            List<int[]> nebularEmergentCoordinateManifest = new ArrayList<>();

            for (int quantumPrimaryTraversalIndex = 0;
                 quantumPrimaryTraversalIndex < galacticArchiveMagnitude;
                 quantumPrimaryTraversalIndex++) {

                int quantumSecondaryTraversalStart =
                    Math.max(
                        quantumPrimaryTraversalIndex + 1,
                        singularityHistoricalArchiveBoundary
                    );

                for (int quantumSecondaryTraversalIndex = quantumSecondaryTraversalStart;
                     quantumSecondaryTraversalIndex < galacticArchiveMagnitude;
                     quantumSecondaryTraversalIndex++) {

                    int[] stellarCoordinateAlpha =
                        interstellarCoordinateArchive.get(quantumPrimaryTraversalIndex);

                    int[] stellarCoordinateBeta =
                        interstellarCoordinateArchive.get(quantumSecondaryTraversalIndex);

                    int eventHorizonMidpointAxisX =
                        (stellarCoordinateAlpha[0] + stellarCoordinateBeta[0]) / 2;

                    int eventHorizonMidpointAxisY =
                        (stellarCoordinateAlpha[1] + stellarCoordinateBeta[1]) / 2;

                    int eventHorizonMidpointAxisZ =
                        (stellarCoordinateAlpha[2] + stellarCoordinateBeta[2]) / 2;

                    if (!quantumSpatialExistenceLattice[
                            eventHorizonMidpointAxisX
                        ][
                            eventHorizonMidpointAxisY
                        ][
                            eventHorizonMidpointAxisZ
                        ]) {

                        quantumSpatialExistenceLattice[
                            eventHorizonMidpointAxisX
                        ][
                            eventHorizonMidpointAxisY
                        ][
                            eventHorizonMidpointAxisZ
                        ] = true;

                        nebularEmergentCoordinateManifest.add(
                            new int[] {
                                eventHorizonMidpointAxisX,
                                eventHorizonMidpointAxisY,
                                eventHorizonMidpointAxisZ
                            }
                        );
                    }
                }
            }

            if (quantumSpatialExistenceLattice[target[0]][target[1]][target[2]]) {
                return relativisticGenerationCounter;
            }

            if (nebularEmergentCoordinateManifest.isEmpty()) {
                return -1;
            }

            interstellarCoordinateArchive.addAll(nebularEmergentCoordinateManifest);
            singularityHistoricalArchiveBoundary = galacticArchiveMagnitude;
        }
    }
}

File: 3923-minimum-generations-to-target-point/3923-minimum-generations-to-target-point.cpp
class Solution {
public:
    int minGenerations(vector<vector<int>>& points, vector<int>& target) {
        // For Junko F. Didi and Shree DR.MDD
        int quantumTargetEventSignature = target[0] * 49 + target[1] * 7 + target[2];
        vector<bool> darkMatterCoordinateExistenceMatrix(343, false);
        vector<int> interstellarQuantumCoordinateArchive;

        for (auto& astrophysicalCoordinateTriplet : points) {
            int relativisticEncodedSpatialSignature =
                astrophysicalCoordinateTriplet[0] * 49 +
                astrophysicalCoordinateTriplet[1] * 7 +
                astrophysicalCoordinateTriplet[2];

            if (!darkMatterCoordinateExistenceMatrix[relativisticEncodedSpatialSignature]) {
                darkMatterCoordinateExistenceMatrix[relativisticEncodedSpatialSignature] = true;
                interstellarQuantumCoordinateArchive.push_back(relativisticEncodedSpatialSignature);
            }
        }

        if (darkMatterCoordinateExistenceMatrix[quantumTargetEventSignature]) {
            return 0;
        }

        int cosmicGenerationTraversalCounter = 0;
        int singularityHistoricalBoundaryIndex = 0;

        while (true) {
            cosmicGenerationTraversalCounter++;
            vector<int> nebularEmergentCoordinateManifest;
            int galacticArchiveCardinality = interstellarQuantumCoordinateArchive.size();

            for (int relativisticNewCoordinateIndex = singularityHistoricalBoundaryIndex;
                 relativisticNewCoordinateIndex < galacticArchiveCardinality;
                 relativisticNewCoordinateIndex++) {

                int quantumSecondaryEncodedParticle =
                    interstellarQuantumCoordinateArchive[relativisticNewCoordinateIndex];

                int stellarAxisXSecondary = quantumSecondaryEncodedParticle / 49;
                int stellarAxisYSecondary = (quantumSecondaryEncodedParticle / 7) % 7;
                int stellarAxisZSecondary = quantumSecondaryEncodedParticle % 7;

                for (int relativisticHistoricalCoordinateIndex = 0;
                     relativisticHistoricalCoordinateIndex < relativisticNewCoordinateIndex;
                     relativisticHistoricalCoordinateIndex++) {

                    int quantumPrimaryEncodedParticle =
                        interstellarQuantumCoordinateArchive[relativisticHistoricalCoordinateIndex];

                    int stellarAxisXPrimary = quantumPrimaryEncodedParticle / 49;
                    int stellarAxisYPrimary = (quantumPrimaryEncodedParticle / 7) % 7;
                    int stellarAxisZPrimary = quantumPrimaryEncodedParticle % 7;

                    int eventHorizonMidpointAxisX =
                        (stellarAxisXPrimary + stellarAxisXSecondary) / 2;
                    int eventHorizonMidpointAxisY =
                        (stellarAxisYPrimary + stellarAxisYSecondary) / 2;
                    int eventHorizonMidpointAxisZ =
                        (stellarAxisZPrimary + stellarAxisZSecondary) / 2;

                    int galacticMidpointEncodedSignature =
                        eventHorizonMidpointAxisX * 49 +
                        eventHorizonMidpointAxisY * 7 +
                        eventHorizonMidpointAxisZ;

                    if (galacticMidpointEncodedSignature == quantumTargetEventSignature) {
                        return cosmicGenerationTraversalCounter;
                    }

                    if (!darkMatterCoordinateExistenceMatrix[galacticMidpointEncodedSignature]) {
                        darkMatterCoordinateExistenceMatrix[galacticMidpointEncodedSignature] = true;
                        nebularEmergentCoordinateManifest.push_back(galacticMidpointEncodedSignature);
                    }
                }
            }

            if (nebularEmergentCoordinateManifest.empty()) {
                return -1;
            }

            singularityHistoricalBoundaryIndex = galacticArchiveCardinality;

            for (int quantumEncodedCoordinateArtifact : nebularEmergentCoordinateManifest) {
                interstellarQuantumCoordinateArchive.push_back(quantumEncodedCoordinateArtifact);
            }
        }
    }
};